src/main/java/io/odpf/firehose/sink/clickhouse/ClickhouseSinkFactory.java

```diff
@@ -0,0 +1,80 @@
+ * using the configurations supplied and invoke {@see #create(Map < String, String > configuration, StatsDReporter statsDReporter, StencilClient client)}
+ * to obtain the Clickhouse sink implementation.
+ */
+public class ClickhouseSinkFactory {
+
+    private final Map<String, String> config;
+    private final StatsDReporter statsDReporter;
+    private final StencilClient stencilClient;
+    private QueryTemplate queryTemplate;
+    private ClickHouseClient client;
+    private ClickHouseNode server;
+    private ClickHouseRequest request;
+
+    /**
+     * Parameterised Constructor to create Clickhouse sink factory object.
+     *
+     * @param config         the configuration
+     * @param statsDReporter the statsd reporter
+     * @param stencilClient  the client
+     */
+
+    public ClickhouseSinkFactory(Map<String, String> config, StatsDReporter statsDReporter, StencilClient stencilClient) {
+        this.config = config;
+        this.statsDReporter = statsDReporter;
+        this.stencilClient = stencilClient;
+    }
+
+    public void init() {
+        ClickhouseSinkConfig clickhouseSinkConfig = ConfigFactory.create(ClickhouseSinkConfig.class, config);
+        server = ClickHouseNode.builder()
+                .host(clickhouseSinkConfig.getClickhouseHost())
+                .port(ClickHouseProtocol.HTTP, Integer.parseInt(clickhouseSinkConfig.getClickhousePort()))
+                .database(clickhouseSinkConfig.getClickhouseDatabase()).credentials(ClickHouseCredentials.fromUserAndPassword(
+                        clickhouseSinkConfig.getClickhouseUsername(), clickhouseSinkConfig.getClickhousePassword()))
+                .build();
+        client = ClickHouseClient.builder()
+                .nodeSelector(ClickHouseNodeSelector.of(server.getProtocol()))
+                .option(ClickHouseClientOption.ASYNC, clickhouseSinkConfig.isClickhouseAsyncModeEnabled())
+                .option(ClickHouseClientOption.COMPRESS_ALGORITHM, clickhouseSinkConfig.getClickhouseCompressionAlgorithm()).option(ClickHouseClientOption.COMPRESS, clickhouseSinkConfig.isClickhouseCompressEnabled())
+                .option(ClickHouseClientOption.DECOMPRESS_ALGORITHM, clickhouseSinkConfig.getClickhouseCompressionAlgorithm()).option(ClickHouseClientOption.DECOMPRESS, clickhouseSinkConfig.isClickhouseDecompressEnabled())
+                .build();
+        Instrumentation instrumentation = new Instrumentation(statsDReporter, ClickhouseSinkFactory.class);
+        instrumentation.logInfo("Clickhouse connection established");
+
+        request = client.connect(server);
+        Parser protoParser = stencilClient.getParser(clickhouseSinkConfig.getInputSchemaProtoClass());
+        ProtoToFieldMapper protoToFieldMapper = new ProtoToFieldMapper(protoParser, clickhouseSinkConfig.getInputSchemaProtoToColumnMapping());
+        queryTemplate = new QueryTemplate(clickhouseSinkConfig, protoToFieldMapper);
```

Comment:
we discussed this that QueryTemplate is not thread safe.. it cannot be common in different sinks.

Reply:
@lavkesh Done.